pipeline-model-api/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/validator/AbstractModelValidator.java

```diff
@@ -0,0 +1,233 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.pipeline.modeldefinition.validator;
+
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTAgent;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTAxis;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTAxisContainer;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTBranch;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTBuildCondition;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTBuildConditionsContainer;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTBuildParameter;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTBuildParameters;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTEnvironment;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTExclude;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTExcludeAxis;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTExcludes;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTInternalFunctionCall;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTLibraries;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTMatrix;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTMethodCall;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTOption;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTOptions;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTParallel;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTPipelineDef;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTPostBuild;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTPostStage;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTStage;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTStageBase;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTStageInput;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTStages;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTStep;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTTools;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTTrigger;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTTriggers;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTValue;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTWhen;
+import org.jenkinsci.plugins.pipeline.modeldefinition.ast.ModelASTWhenCondition;
+
+/**
+ * Abstract implementation of {@link ModelValidator}.
+ *
+ * Use this class as a generic AST visitor instead of {@link ModelValidator} to prevent binary compatibility issues
```

Comment:
Maybe also good to update {{ModelValidator}}'s Javadoc to explicitly note that it is not a stable API and should not be used directly by plugins other than Declarative itself?